### DISCUSSION TITLE
Some questions about Ziffers....

### DISCUSSION THREAD

**Discussion by beryann** [amiika/ziffers#86] [General]
Hello Amiika! as I wrote it on discord, here are a first serie of questions I noted learning the Ziffers notation: 

Questions about the Ziffers cheatsheet :

- zplay "{2*2} {3*3}" # Eval math  What does it do exactly ? 4 and 9 okay but are they notes ?

- zplay "[4 2 4 2] [4 5 [4 2]] [3 [1 [3 1]]]" : are there 3 beats ?
              1             2               3                                                                                                                                      

- zplay "0 1 2 3", rhythm: {major: 7, minor: 3} : I don’t understand this one

- zplay "q (X S)<2,4>", X: :bd_haus, S: :bd_zum : there is a bug  inthe log (« Timing error: can't keep up »

- z1 "((e 4 2) (e 0 ?))<6,8>(q6 q(5,7))"  what does exactly the loop ?

-zplay "0 1 2 3", rhythm: {major: 7, minor: 3} How doest it work ?

pattern rotation: zplay "q (X S)<6,9,3> supposed S : bd_zum f.i ?

-zplay "i 1 2 3 ii vi v 026", chord_key: :D or E or C doesn’t change the notes in the log when I print them ? I mean, that « chord_key ::d or e » doesn’t affect the notes when I hear it…

-zplay "ii%-2", here the minus sign changes the octave ? And « 2 » is I suppose the second inversion

-here : « iv^maj9%-4" » the e in classical harmony is supposed to be eb, is it he case here> ? And once more, what does the '-' ?

-z1 "([: i^maj :3] [: iv^dim7%-1 :3])@(q 0 1 e 2 1 0 1)",
  key: :d4, scale: :major

-z1 "([: i^maj :3] [: iv^dim7%-1 :3])@(q 0 1 e 2 1 0 1)",
  key: :d4, scale: :major : here the dim7 chord for its 1 inversion is supposed to be on the 4th degree in D : Bb, Cs, E, G and the sequence is E, G, A

-z1 "0 1 2 3", add: [-1,2,0,2] : what does exactly the array ?

Thanks a lot for your help! :-))

**Comment by amiika** [comment]
Eval (or curly braces notation) evaluates math or espaces multiple pitches (since by default multiple consecutive pitches are a chord in default ziffers notation), so for example: {8 9 10 11 12 3*5} plays 8, 9, 10, 11, 12, 15 pitches from the scale.

**Comment by amiika** [comment]
"[4 2 4 2] [4 5 [4 2]] [3 [1 [3 1]]]" subdivides note length, which by default is a whole note = 4 beats. Try changing the note length before the subdivision for example "q [4 2 4 2] 0.5 [4 5 [4 2]] 1/8 [3 [1 [3 1]]]"

**Comment by amiika** [comment]
Minor and major rhythm generators are ideas based on Schillingers rhythm theory which is bit obscure ... major and minor refers to generators of pulses going out of synch with each other, like three against four or four against five and so on. It was just an experiment to learn it from the Schillinger system book which can be found [from internet archive](https://ia801008.us.archive.org/30/items/SchillingerSystem/Schillinger%20System%20of%20Musical%20Composition_text.pdf).

Don't know if it's worth it tho, deep hole that one is :D

**Comment by amiika** [comment]
Timing error: can't keep up -> Means Sonic Pi is pushing the limits of the computer / allocated resources ... can be caused by many things. Be sure to require Ziffers in the init file only once ... use require and not load ... and allocate as much juice to Sonic Pi as possible.

**Comment by amiika** [comment]
Chords inversions go up or down, for example "ii%-2" is two down.

For the rest of the chord examples I have to check later when I have Sonic Pi running again :)

**Comment by amiika** [comment]
z1 "0 1 2 3", add: [-1,2,0,2]  -> I think this does addition picking next one from the array once for the whole loop. Alternatively you can try changing array to ring; add: [-1,2,0,2].ring which then cycles it for every pitch in the sequence.

**Comment by amiika** [comment]
Support for chord_key seems to be dropped out of the code at some point, ill make issue out of this one

**Comment by amiika** [comment]
z1 "((e 4 2) (e 0 ?))<6,8>(q6 q(5,7))" is [ziffers euclidean notation](z1 "((e 4 2) (e 0 ?))<6,8>(q6 q(5,7))"): (onbeat cycle)<6,8>(offbeat cycle) ... If there are multiple patterns inside the on/offbeat cycles it just treats em like cycles.

Try simple examples first ... something like:

z1 "(0)<5,7>(4)" vs. z1 "((q 0) (e 0 ?))<5,7>(4)"
